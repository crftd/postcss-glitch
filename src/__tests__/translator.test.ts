/**
 * Crafted by Crash on 29.11.17.
 */

import { decl, Declaration, root, Root, rule, Rule } from 'postcss';
import Translator from '../translator';
import mockClipPath from '../clip-path.builder';

jest.mock('../clip-path.builder');

describe('translator', () => {
  const originalTranslator = {
    addPseudo: Translator.addPseudo,
    addKeyframes: Translator.addKeyframes,
    removeDeclaration: Translator.removeDeclaration,
  };
  const expectedSelector = '.foo';
  const expectedHeight = '48px';
  const expectedShadowOffset = '1px';
  const expectedFirstColor = '#f00';
  const expectedSecondColor = '#00f';

  let expectedRoot: Root;
  let expectedRule: Rule;

  beforeEach(() => {
    expectedRoot = root();
    expectedRule = rule({ selector: expectedSelector });
  });

  afterEach(() => {
    Translator.addPseudo = originalTranslator.addPseudo;
    Translator.addKeyframes = originalTranslator.addKeyframes;
    Translator.removeDeclaration = originalTranslator.removeDeclaration;
  });

  it('translate declaration', () => {
    // Arrange
    const expectedDeclaration = decl({
      prop: 'glitch',
      value: `${expectedHeight} ${expectedFirstColor} ${expectedSecondColor} ${expectedShadowOffset}`,
    });

    const mockAddPseudo = jest.fn();
    const mockAddKeyframes = jest.fn();
    const mockRemoveDeclaration = jest.fn();

    Translator.addPseudo = mockAddPseudo;
    Translator.addKeyframes = mockAddKeyframes;
    Translator.removeDeclaration = mockRemoveDeclaration;

    // Act
    Translator.translate(expectedDeclaration);

    // Assert
    expect(mockAddPseudo).toHaveBeenCalledWith(expectedDeclaration);
    expect(mockAddKeyframes).toHaveBeenCalledWith(expectedDeclaration);
    expect(mockRemoveDeclaration).toHaveBeenCalledWith(expectedDeclaration);
  });

  it('adds ::before and ::after pseudo elements to the rule that contains glitch declaration', () => {
    // Arrange
    const expectedDeclaration = decl({
      prop: 'glitch',
      value: `${expectedHeight} ${expectedFirstColor} ${expectedSecondColor} ${expectedShadowOffset}`,
    });
    expectedRule.append(expectedDeclaration);
    expectedRoot.append(expectedRule);
    const expectedResultCss = `${expectedSelector} {
    glitch: ${expectedHeight} ${expectedFirstColor} ${expectedSecondColor} ${expectedShadowOffset}
}
${expectedSelector}::before, ${expectedSelector}::after {
    content: attr(data-text);
    position: absolute;
    top: 0;
    left: 0;
    overflow: hidden;
    clip-path: inset(${expectedHeight} 0 0 0)
}
${expectedSelector}::before {
    text-shadow: -${expectedShadowOffset} 0 ${expectedFirstColor};
    animation: glitch-animation-before alternate-reverse 3s infinite linear
}
${expectedSelector}::after {
    text-shadow: ${expectedShadowOffset} 0 ${expectedSecondColor};
    animation: glitch-animation-after alternate-reverse 2s infinite linear
}`;

    // Act
    Translator.addPseudo(expectedDeclaration);

    // Assert
    expect(expectedRoot.toString()).toEqual(expectedResultCss);
  });

  it('adds 2 @keyframes animation with 21 steps at root and glitch height 5px', () => {
    // Arrange
    const expectedOffsetTop = '24px';
    const expectedOffsetBottom = '19px';
    const expectedClipPath = decl({
      prop: 'clip-path',
      value: `inset(${expectedOffsetTop} 0 ${expectedOffsetBottom} 0)`,
    });
    const expectedDeclaration = decl({
      prop: 'glitch',
      value: `${expectedHeight} ${expectedFirstColor} ${expectedSecondColor} ${expectedShadowOffset}`,
    });
    expectedRule.append(expectedDeclaration);
    expectedRoot.append(expectedRule);
    (mockClipPath as jest.Mock).mockImplementation((): Declaration => expectedClipPath.clone());

    // Act
    Translator.addKeyframes(expectedDeclaration);

    // Assert
    expect(expectedRoot.toString()).toMatchSnapshot();
  });

  it('removes glitch declaration', () => {
    // Arrange
    const expectedDeclaration = decl({
      prop: 'glitch',
      value: `${expectedHeight} ${expectedFirstColor} ${expectedSecondColor} ${expectedShadowOffset}`,
    });
    expectedRule.append(expectedDeclaration);
    expectedRoot.append(expectedRule);
    const expectedResultCss = `${expectedSelector} {}`;

    // Act
    Translator.removeDeclaration(expectedDeclaration);

    // Assert
    expect(expectedRoot.toString()).toEqual(expectedResultCss);
  });
});
