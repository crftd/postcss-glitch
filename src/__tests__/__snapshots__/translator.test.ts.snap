// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translator adds 2 @keyframes animation with 21 steps at root and glitch height 5px 1`] = `
"@keyframes glitch-animation-before {
    0% {
        clip-path: inset(24px 0 19px 0)
    }
    5% {
        clip-path: inset(24px 0 19px 0)
    }
    10% {
        clip-path: inset(24px 0 19px 0)
    }
    15% {
        clip-path: inset(24px 0 19px 0)
    }
    20% {
        clip-path: inset(24px 0 19px 0)
    }
    25% {
        clip-path: inset(24px 0 19px 0)
    }
    30% {
        clip-path: inset(24px 0 19px 0)
    }
    35% {
        clip-path: inset(24px 0 19px 0)
    }
    40% {
        clip-path: inset(24px 0 19px 0)
    }
    45% {
        clip-path: inset(24px 0 19px 0)
    }
    50% {
        clip-path: inset(24px 0 19px 0)
    }
    55% {
        clip-path: inset(24px 0 19px 0)
    }
    60% {
        clip-path: inset(24px 0 19px 0)
    }
    65% {
        clip-path: inset(24px 0 19px 0)
    }
    70% {
        clip-path: inset(24px 0 19px 0)
    }
    75% {
        clip-path: inset(24px 0 19px 0)
    }
    80% {
        clip-path: inset(24px 0 19px 0)
    }
    85% {
        clip-path: inset(24px 0 19px 0)
    }
    90% {
        clip-path: inset(24px 0 19px 0)
    }
    95% {
        clip-path: inset(24px 0 19px 0)
    }
    100% {
        clip-path: inset(24px 0 19px 0)
    }
}
@keyframes glitch-animation-after {
    0% {
        clip-path: inset(24px 0 19px 0)
    }
    5% {
        clip-path: inset(24px 0 19px 0)
    }
    10% {
        clip-path: inset(24px 0 19px 0)
    }
    15% {
        clip-path: inset(24px 0 19px 0)
    }
    20% {
        clip-path: inset(24px 0 19px 0)
    }
    25% {
        clip-path: inset(24px 0 19px 0)
    }
    30% {
        clip-path: inset(24px 0 19px 0)
    }
    35% {
        clip-path: inset(24px 0 19px 0)
    }
    40% {
        clip-path: inset(24px 0 19px 0)
    }
    45% {
        clip-path: inset(24px 0 19px 0)
    }
    50% {
        clip-path: inset(24px 0 19px 0)
    }
    55% {
        clip-path: inset(24px 0 19px 0)
    }
    60% {
        clip-path: inset(24px 0 19px 0)
    }
    65% {
        clip-path: inset(24px 0 19px 0)
    }
    70% {
        clip-path: inset(24px 0 19px 0)
    }
    75% {
        clip-path: inset(24px 0 19px 0)
    }
    80% {
        clip-path: inset(24px 0 19px 0)
    }
    85% {
        clip-path: inset(24px 0 19px 0)
    }
    90% {
        clip-path: inset(24px 0 19px 0)
    }
    95% {
        clip-path: inset(24px 0 19px 0)
    }
    100% {
        clip-path: inset(24px 0 19px 0)
    }
}
.foo {
    glitch: 48px #f00 #00f 1px
}"
`;
